liota/lib/transports/mqtt.py

```diff
@@ -320,12 +323,27 @@ def subscribe(self, topic, qos, callback):
         :return:
         """
         try:
+            if [topic,qos,callback] not in self.sub_list:
```

Comment:
In what situations would we try to add an additional subscription for the same thing? And what happens if we subscribe to the same topic with a different qos or callback? Given unsubscribe only cares about the topic, I'm assuming whatever we subscribe with should only care about the topic, too? Or otherwise overwrite the old qos and callback?

Reply:
I did some testing and figured out that the Paho Python mqtt client subscribes to only one topic and basically it should be overriding the QoS and callback if the new subscription arrives for the same topic as I didn't receive duplicate messages. This check I have included keeping that in mind as some desktop mqtt client allow to subscribe to multiple topics and the messages get duplicated.

Reply:
Hm. But then sub_list is not an accurate representation of what is subscribed, right?

Subscribe(foo, qos1, myCallback)
Subscribe(foo, qos2, myOtherCallback)
Unsubscribe(foo)

What's in the list and what is the client subscribed to now?

Reply:
I have already opened an issue at Paho Python MQTT Github project to understand the behavior correctly:

https://github.com/eclipse/paho.mqtt.python/issues/241

As per my testing as library subscribes to only one topic with QoS & Callback, I plan to move from sub_list to python dictionary where the key will be the topic & values will be [Qos, Callback] and if the topic is unsubscribed it will be removed from the map. I`ll wait for a day for the final confirmation from the paho developers on the library behavior before making this change.

Reply:
In message_callback_add(), Paho MQTT client is adding the new callback and doesn't keep a list of all callbacks registered till now. (https://github.com/eclipse/paho.mqtt.python/blob/master/src/paho/mqtt/client.py#L1735)

I think it is better to remove the check.

Reply:
Implemented a python dictionary rather than list it adds the topic,QoS & Callback on_subscribe() method and removes them on_unsubscribe()